lib/dao.js

```diff
@@ -624,7 +624,197 @@ DataAccessObject.upsert = function(data, options, cb) {
+        var doValidate = undefined;
+        if (options.validate === undefined) {
+          if (Model.settings.validateUpsert === undefined) {
+            if (Model.settings.automaticValidation !== undefined) {
+              doValidate = Model.settings.automaticValidation;
+            }
+          } else {
+            doValidate = Model.settings.validateUpsert;
+          }
+        } else {
+          doValidate = options.validate;
+        }
+        if (doValidate === false) {
+          callConnector();
+        } else {
+          inst.isValid(function(valid) {
+            if (!valid) {
+              console.warn('Ignoring validation errors in updateOrCreate():');
+              console.warn('  %s', new ValidationError(inst).message);
+            }
+            callConnector();
+          }, update, options);
+        }
+        function callConnector() {
+          update = removeUndefined(update);
+          context = {
+            Model: Model,
+            where: ctx.where,
+            data: update,
+            currentInstance: inst,
+            hookState: ctx.hookState,
+            options: options,
+          };
+          Model.notifyObserversOf('persist', context, function(err) {
+            if (err) return done(err);
+            connector.upsertWithWhere(modelName, where, update, options, done);
+          });
+        }
+        function done(err, data, info) {
+          if (err) return cb(err);
+          var context = {
+            Model: Model,
+            data: data,
+            isNewInstance: info && info.isNewInstance,
+            hookState: ctx.hookState,
+            options: options,
+          };
+          Model.notifyObserversOf('loaded', context, function(err) {
+            if (err) return cb(err);
+            var obj;
+            if (data && !(data instanceof Model)) {
+              inst._initProperties(data, { persisted: true });
+              obj = inst;
+            } else {
+              obj = data;
+            }
+            if (err) {
+              cb(err, obj);
+            } else {
+              var context = {
+                Model: Model,
+                instance: obj,
+                isNewInstance: info ? info.isNewInstance : undefined,
+                hookState: hookState,
+                options: options,
+              };
+              Model.notifyObserversOf('after save', context, function(err) {
+                cb(err, obj);
+              });
+            }
+          });
+        }
+      });
+    } else {
+      var element = Object.keys(where)[0];
+      var filter = { 'where': where };
+      if (!Object.keys(data).length) {
+        var err = new Error('data object cannot be empty!');
+        err.statusCode = 400;
+        process.nextTick(function() { cb(err); });
+        return cb.promise;
+      }
 
+      if (!data.hasOwnProperty(element)) {
+        data[element] = where[element];
+      }
+      var pk = idName(Model);
```

Comment:
@superkhau  That variable won;t be needed now, after removing the data.hasOwnProperty(pk) === false check in the previous comment from @Amir-61 . Thanks for pointing it out.